tests/js-custom-authorization/custom_authorization.py

```diff
@@ -356,6 +365,339 @@ def test_jwt_auth(network, args):
+                        "x5c": [der_b64],
+                        "issuer": ISSUER_TENANT,
+                    },
+                ]
+            },
+        }
+        json.dump(data, metadata_fp)
+        metadata_fp.flush()
+        network.consortium.set_jwt_issuer(primary, metadata_fp.name)
+
+    assert (
+        try_auth(primary, issuer, jwt_kid_1, ISSUER_TENANT, TENANT_ID) == HTTPStatus.OK
+    )
+    assert (
+        try_auth(primary, issuer, jwt_kid_1, ISSUER_ANOTHER, ANOTHER_TENANT_ID)
+        == HTTPStatus.OK
+    )
+    assert (
+        try_auth(primary, issuer, jwt_kid_1, ISSUER_TENANT, ANOTHER_TENANT_ID)
+        == HTTPStatus.UNAUTHORIZED
+    )
+
+    assert (
+        try_auth(primary, issuer, jwt_kid_2, ISSUER_TENANT, TENANT_ID) == HTTPStatus.OK
+    )
+    assert (
+        try_auth(primary, issuer, jwt_kid_2, ISSUER_ANOTHER, ANOTHER_TENANT_ID)
+        == HTTPStatus.UNAUTHORIZED
+    )
+
+    network.consortium.remove_jwt_issuer(primary, issuer.name)
+
+    assert (
+        try_auth(primary, issuer, jwt_kid_1, ISSUER_TENANT, TENANT_ID)
+        == HTTPStatus.UNAUTHORIZED
+    )
+    assert (
+        try_auth(primary, issuer, jwt_kid_2, ISSUER_TENANT, TENANT_ID)
+        == HTTPStatus.UNAUTHORIZED
+    )
+
+    return network
+
+
+@reqs.description("JWT authentication with same kids for different issuers")
+def test_jwt_auth_msft_same_kids_different_issuers(network, args):
+    """Multiple issuer can share same kid with their own constraints specified.
+    So when issuer1 adds it with constraint1, and issuer2 added it with constraint2,
+    then tokens from both issuer1 and issuer2 can go.
+    However, when issuer1 is removed, issuer2 has to remain capable of
+    authenticating with their tokens."""
+
+    primary, _ = network.find_nodes()
+
+    TENANT_ID = "9188050d-6c67-4c5b-b112-36a304b66da"
+    ISSUER_TENANT = f"https://login.microsoftonline.com/{TENANT_ID}/v2.0"
+    ANOTHER_TENANT_ID = "ANOTHER-6c67-4c5b-b112-36a304b66da"
+    ISSUER_ANOTHER = f"https://login.microsoftonline.com/{ANOTHER_TENANT_ID}/v2.0"
+
+    issuer = infra.jwt_issuer.JwtIssuer(name=ISSUER_TENANT)
+    another = infra.jwt_issuer.JwtIssuer(name=ISSUER_ANOTHER)
+
+    jwt_kid = "my_key_id"
+
+    with tempfile.NamedTemporaryFile(prefix="ccf", mode="w+") as metadata_fp:
+        jwt_cert_der = infra.crypto.cert_pem_to_der(issuer.cert_pem)
+        der_b64 = base64.b64encode(jwt_cert_der).decode("ascii")
+        data = {
+            "issuer": issuer.issuer_url,
+            "auto_refresh": False,
+            "jwks": {
+                "keys": [
+                    {
+                        "kty": "RSA",
+                        "kid": jwt_kid,
+                        "x5c": [der_b64],
+                        "issuer": ISSUER_TENANT,
+                    }
+                ]
+            },
+        }
+        json.dump(data, metadata_fp)
+        metadata_fp.flush()
+        network.consortium.set_jwt_issuer(primary, metadata_fp.name)
```

Comment:
We do this block several times - can we pull it out to a helper function?

Reply:
Done